content_script.js

```diff
@@ -137,38 +140,104 @@ class HTMLSerializer {
    *
    * @param {Element} element The element being processed, which has the src
    *     attribute.
+   * @param {Set<string>} processedAttributes The Set containing all attributes
+   *     already added to |this.html| for the Element.
    * @private
    */
-  processSrcAttribute(element) {
-    if (element.tagName.toLowerCase() != 'iframe') {
-      this.processSrcHole(element);
-    } 
+  processSrcAttribute(element, processedAttributes) {
+    var tag = element.tagName;
+    switch (tag) {
+      case 'IFRAME':
+        break; // Do nothing.
+      case 'SOURCE':
+        if (!element.parent || element.parent.tagName != 'IMG') {
+          this.processSimpleSrc(element, processedAttributes);
```

Comment:
Why does IMG need the location.host check but this doesn't? Can the location.host logic be moved into processSimpleSrc?

It's confusing how SOURCE falls through INPUT to IMG to default. Instead of falling through, what do you think of explicitly handling the code in each case?

Can you add a test of source with a parent of img? I thought this was not allowed, but https://developer.mozilla.org/en-US/docs/Web/HTML/Element/source isn't very clear on that.